File: SCT2/trunk/test-plugins/org.yakindu.sct.generator.cpp.test/gtests/SyncJoinTest/SyncJoinTest.cc
/**
* Copyright (c) 2013 committers of YAKINDU and others.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Eclipse Public License v1.0
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v10.html
*
* Contributors:
*     committers of YAKINDU - initial API and implementation
*/
#include <string>
#include "gtest/gtest.h"
#include "SyncJoin.h"

TEST(StatemachineTest, syncJoin_C2_Waits) {
	SyncJoin* statechart = new SyncJoin();
	statechart->init();
	statechart->enter();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r1_C1));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r2_D1));
	statechart->raise_e();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r1_C2));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r2_D1));
	statechart->raise_jc();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r1_C2));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r2_D1));
	statechart->raise_jd();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r1_C2));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r2_D1));
	statechart->raise_jc();
	statechart->raise_jd();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r1_C2));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r2_D1));
	statechart->raise_f();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r1_C2));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r2_D2));
	statechart->raise_jc();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r1_C2));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r2_D2));
	statechart->raise_jd();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r1_C2));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r2_D2));
	statechart->raise_jc();
	statechart->raise_jd();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_A));
	delete statechart;
}
TEST(StatemachineTest, syncJoin_D2_Waits) {
	SyncJoin* statechart = new SyncJoin();
	statechart->init();
	statechart->enter();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r1_C1));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r2_D1));
	statechart->raise_f();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r1_C1));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r2_D2));
	statechart->raise_jc();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r1_C1));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r2_D2));
	statechart->raise_jd();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r1_C1));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r2_D2));
	statechart->raise_jc();
	statechart->raise_jd();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r1_C1));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r2_D2));
	statechart->raise_e();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r1_C2));
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_B_r2_D2));
	delete statechart;
}
TEST(StatemachineTest, doubleEntryActionBug) {
	SyncJoin* statechart = new SyncJoin();
	statechart->init();
	statechart->enter();
	statechart->raise_e();
	statechart->raise_f();
	statechart->runCycle();
	statechart->raise_jc();
	statechart->raise_jd();
	statechart->runCycle();
	EXPECT_TRUE(statechart->isActive(SyncJoin::SyncJoin_main_region_A));
	EXPECT_TRUE(statechart->getSCInterface()->get_x()== 1);
	delete statechart;
}

File: test-plugins/org.yakindu.sct.simulation.core.sexec.test/test-gen/org/yakindu/sct/simulation/core/sexec/test/SyncJoinTest.java
/**
 * Copyright (c) 2015 committers of YAKINDU and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     committers of YAKINDU - initial API and implementation
 */
package org.yakindu.sct.simulation.core.sexec.test;
import org.eclipse.xtext.junit4.InjectWith;
import org.eclipse.xtext.junit4.XtextRunner;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.yakindu.sct.model.sexec.ExecutionFlow;
import org.yakindu.sct.model.sexec.interpreter.test.util.AbstractExecutionFlowTest;
import org.yakindu.sct.model.sexec.interpreter.test.util.SExecInjectionProvider;
import org.yakindu.sct.test.models.SCTUnitTestModels;
import com.google.inject.Inject;
import static org.junit.Assert.assertTrue;
/**
 *  Unit TestCase for SyncJoin
 */
@SuppressWarnings("all")
@RunWith(XtextRunner.class)
@InjectWith(SExecInjectionProvider.class)
public class SyncJoinTest extends AbstractExecutionFlowTest {
	@Before
	public void setup() throws Exception {
		ExecutionFlow flow = models
				.loadExecutionFlowFromResource("SyncJoin.sct");
		initInterpreter(flow);
	}
	@Test
	public void syncJoin_C2_Waits() throws Exception {
		interpreter.enter();
		assertTrue(isStateActive("B"));
		assertTrue(isStateActive("C1"));
		assertTrue(isStateActive("D1"));
		raiseEvent("e");
		interpreter.runCycle();
		assertTrue(isStateActive("C2"));
		assertTrue(isStateActive("D1"));
		raiseEvent("jc");
		interpreter.runCycle();
		assertTrue(isStateActive("C2"));
		assertTrue(isStateActive("D1"));
		raiseEvent("jd");
		interpreter.runCycle();
		assertTrue(isStateActive("C2"));
		assertTrue(isStateActive("D1"));
		raiseEvent("jc");
		raiseEvent("jd");
		interpreter.runCycle();
		assertTrue(isStateActive("C2"));
		assertTrue(isStateActive("D1"));
		raiseEvent("f");
		interpreter.runCycle();
		assertTrue(isStateActive("C2"));
		assertTrue(isStateActive("D2"));
		raiseEvent("jc");
		interpreter.runCycle();
		assertTrue(isStateActive("C2"));
		assertTrue(isStateActive("D2"));
		raiseEvent("jd");
		interpreter.runCycle();
		assertTrue(isStateActive("C2"));
		assertTrue(isStateActive("D2"));
		raiseEvent("jc");
		raiseEvent("jd");
		interpreter.runCycle();
		assertTrue(isStateActive("A"));
	}
	@Test
	public void syncJoin_D2_Waits() throws Exception {
		interpreter.enter();
		assertTrue(isStateActive("B"));
		assertTrue(isStateActive("C1"));
		assertTrue(isStateActive("D1"));
		raiseEvent("f");
		interpreter.runCycle();
		assertTrue(isStateActive("C1"));
		assertTrue(isStateActive("D2"));
		raiseEvent("jc");
		interpreter.runCycle();
		assertTrue(isStateActive("C1"));
		assertTrue(isStateActive("D2"));
		raiseEvent("jd");
		interpreter.runCycle();
		assertTrue(isStateActive("C1"));
		assertTrue(isStateActive("D2"));
		raiseEvent("jc");
		raiseEvent("jd");
		interpreter.runCycle();
		assertTrue(isStateActive("C1"));
		assertTrue(isStateActive("D2"));
		raiseEvent("e");
		interpreter.runCycle();
		assertTrue(isStateActive("C2"));
		assertTrue(isStateActive("D2"));
	}
	@Test
	public void doubleEntryActionBug() throws Exception {
		interpreter.enter();
		raiseEvent("e");
		raiseEvent("f");
		interpreter.runCycle();
		raiseEvent("jc");
		raiseEvent("jd");
		interpreter.runCycle();
		assertTrue(isStateActive("A"));
		assertTrue(getInteger("x") == 1l);
	}
}
